ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/ToolsFactory.java

```diff
@@ -0,0 +1,37 @@
+package org.opensearch.ml.engine.tools;
+
+import org.opensearch.client.Client;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.ml.common.spi.tools.Tool;
+import org.opensearch.script.ScriptService;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class ToolsFactory {
+    private static final Map<String, Tool> AGENT_TOOLS = new HashMap<>();
+
+    public ToolsFactory(Client client, ScriptService scriptService, ClusterService clusterService,
+                        NamedXContentRegistry xContentRegistry, Map<String, Tool> externalTools) {
+        AGENT_TOOLS.put(MLModelTool.TYPE, new MLModelTool(client));
```

Comment:
Do we only have 1 instance for each tool?

Reply:
This depends upon tool implementation. For example, VectorDBTool was not thread safe but have updated the tool as part of this CR to be thread safe.

Reply:
It's better mentioned in comments, each plugin need to be thread safe.

Reply:
Added documentation to Tool and Abstract tool interfaces.

Reply:
Before this refactoring, we allow user to define multiple Tools with same type. For example, user can define tools like this
```
[
{
      "type": "VectorDBTool",
      "name": "opensearch_vectordb",
},
{
      "type": "VectorDBTool",
      "name": "aws_vectordb",
}
]
```